### ISSUE TITLE
Metrics compute per batch and not per sample

### ISSUE THREAD

**Issue by diane-wagner** [PyTorchLightning/pytorch-lightning#3572]
## 🐛 Bug

Some metrics (e.g. f1-score) compute per batch and not per sample (e.g. per image). What most people expect is that the metric is computed for each sample and then aggregated. The current behavior is not well documented,  which can lead to potential errors in evaluation.

#### Code sample

```python
>>> import torch
>>> from pytorch_lightning.metrics.functional import f1_score
>>> pred = torch.tensor([[0, 1], [0, 1]])
>>> target = torch.tensor([[0, 1], [0, 0]])
>>> f1_per_batch = f1_score(pred=pred, target=target, num_classes=2, class_reduction="none")

>>> f1_sample_0 = f1_score(pred=pred[0], target=target[0], num_classes=2, class_reduction="none")
>>> f1_sample_1 = f1_score(pred=pred[1], target=target[1], num_classes=2, class_reduction="none")
>>> f1_mean_across_samples = (f1_sample_0 + f1_sample_1) / 2

>>> f1_per_batch
tensor([0.8000, 0.6667])

>>> f1_mean_across_samples
tensor([0.8333, 0.5000])
```
However, `f1_per_batch != f1_mean_across_samples`.

**Comment by edenlightning** [comment]
thank you! We are looking into to this is other issues.
